File: C++/131. Palindrome Partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<string> subS;
        vector<vector<string>> res;
        solve(s,0,subS,res);
        return res;
    }
    void solve(string& s, int idx,vector<string>& subS,vector<vector<string>>& res){
        if (idx==s.size()){
            res.push_back(subS);
            return;
        }
        for (int i = idx; i <s.length() ; i++){
            if(check(s,idx,i)){
                subS.push_back(s.substr(idx,i-idx+1));
                solve(s,i+1,subS,res);
                subS.pop_back();
            }
        }
    }
    bool check(string s, int i, int j){
        while (i<=j){
            if(s[i++]!=s[j--]){
                return false;
            }
        }
        return true;
    }
};

File: Java/131. Palindrome Partitioning.java
class Solution {
    public List<List<String>> partition(String s) {
        List<String> subS=new ArrayList<>();
        List<List<String>> res=new ArrayList<>();
        solve(s,0,subS,res);
        return res;
    }
    void solve(String s, int idx,List<String> subS,List<List<String>> res){
        if (idx==s.length()){
            res.add(new ArrayList<>(subS));
            return;
        }
        for (int i = idx; i <s.length() ; i++){
            if(check(s,idx,i)==true){
                subS.add(s.substring(idx,i+1));
                solve(s,i+1,subS,res);
                subS.remove(subS.size()-1);
            }
        }
    }
    boolean check(String s, int i, int j){
        while (i<=j){
            if(s.charAt(i++) !=s.charAt(j--)){
                return false;
            }
        }
        return true;
    }
}
